encoding/ccf/encode.go

```diff
@@ -0,0 +1,1766 @@
+		err := e.encodeValue(pair.Key, staticKeyType, tids)
+		if err != nil {
+			return nil, err
+		}
+
+		// Get encoded key length (must flush first).
+		e.enc.Flush()
+		keyLength := buf.Len() - off
+
+		// Encode dictionary value as value.
+		err = e.encodeValue(pair.Value, staticElementType, tids)
+		if err != nil {
+			return nil, err
+		}
+
+		// Get encoded pair length (must flush first).
+		e.enc.Flush()
+		pairLength := buf.Len() - off
+
+		encodedPairs[i] = encodedKeyValuePair{keyLength: keyLength, pairLength: pairLength}
+	}
+
+	// Reslice buf for encoded key and pair by offset and length.
+	b := buf.Bytes()
+	off := 0
+	for i := 0; i < len(encodedPairs); i++ {
+		encodedPairs[i].encodedKey = b[off : off+encodedPairs[i].keyLength]
+		encodedPairs[i].encodedPair = b[off : off+encodedPairs[i].pairLength]
+		off += encodedPairs[i].pairLength
+	}
+	if off != len(b) {
+		// Sanity check
+		panic(fmt.Errorf("encoded dictionary pairs' offset %d doesn't match buffer length %d", off, len(b)))
+	}
+
+	sort.Sort(bytewiseKeyValuePairSorter(encodedPairs))
+
+	return encodedPairs, nil
+}
+
+// encodeStruct encodes cadence.Struct as
+// language=CDDL
+// composite-value = [+ (field: value)]
+func (e *Encoder) encodeStruct(v cadence.Struct, tids ccfTypeIDByCadenceType) error {
+	return e.encodeComposite(v.StructType, v.Fields, tids)
+}
+
+// encodeResource encodes cadence.Resource as
+// language=CDDL
+// composite-value = [+ (field: value)]
+func (e *Encoder) encodeResource(v cadence.Resource, tids ccfTypeIDByCadenceType) error {
+	return e.encodeComposite(v.ResourceType, v.Fields, tids)
+}
+
+// encodeEvent encodes cadence.Event as
+// language=CDDL
+// composite-value = [+ (field: value)]
+func (e *Encoder) encodeEvent(v cadence.Event, tids ccfTypeIDByCadenceType) error {
+	return e.encodeComposite(v.EventType, v.Fields, tids)
+}
+
+// encodeContract encodes cadence.Contract as
+// language=CDDL
+// composite-value = [+ (field: value)]
+func (e *Encoder) encodeContract(v cadence.Contract, tids ccfTypeIDByCadenceType) error {
+	return e.encodeComposite(v.ContractType, v.Fields, tids)
+}
+
+// encodeEnum encodes cadence.Enum as
+// language=CDDL
+// composite-value = [+ (field: value)]
+func (e *Encoder) encodeEnum(v cadence.Enum, tids ccfTypeIDByCadenceType) error {
+	return e.encodeComposite(v.EnumType, v.Fields, tids)
+}
+
+// encodeComposite encodes composite types as
+// language=CDDL
+// composite-value = [+ (field: value)]
+func (e *Encoder) encodeComposite(
+	typ cadence.CompositeType,
+	fields []cadence.Value,
+	tids ccfTypeIDByCadenceType,
+) error {
+	staticFieldTypes := typ.CompositeFields()
+
+	if len(staticFieldTypes) != len(fields) {
+		panic(fmt.Errorf(
+			"%s field count %d doesn't match declared field type count %d",
+			typ.ID(),
+			len(fields),
+			len(staticFieldTypes),
+		))
+	}
+
+	// Encode array head with number of fields.
+	err := e.enc.EncodeArrayHead(uint64(len(fields)))
+	if err != nil {
+		return err
+	}
+
```

Comment:
Maybe also short-circuit for `len(fields) == 0`, i.e. just return. Maybe just `switch` on `len(fields)`

Reply:
Done in commit 266fb6a280996e376da1e56ac228bd006ee5c2b0 to match today's change to CCF specs to allow "zero or more" (was "one or more").